Spigot-API-Patches/0304-Add-drops-to-shear-events.patch

```diff
@@ -0,0 +1,107 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Jake Potrebic <jake.m.potrebic@gmail.com>
+Date: Tue, 18 May 2021 12:31:54 -0700
+Subject: [PATCH] Add drops to shear events
+
+
+diff --git a/src/main/java/org/bukkit/event/block/BlockShearEntityEvent.java b/src/main/java/org/bukkit/event/block/BlockShearEntityEvent.java
+index 71c0af9373069cfaa074e1fbad592eab81025b1c..4957a71675b85c854ca368053c8f571157f54002 100644
+--- a/src/main/java/org/bukkit/event/block/BlockShearEntityEvent.java
++++ b/src/main/java/org/bukkit/event/block/BlockShearEntityEvent.java
+@@ -17,11 +17,19 @@ public class BlockShearEntityEvent extends BlockEvent implements Cancellable {
+     private final Entity sheared;
+     private final ItemStack tool;
+     private boolean cancelled;
++    private java.util.List<ItemStack> drops; // Paper
+ 
++    @Deprecated // Paper
+     public BlockShearEntityEvent(@NotNull Block dispenser, @NotNull Entity sheared, @NotNull ItemStack tool) {
++        // Paper start
++        this(dispenser, sheared, tool, com.google.common.collect.Lists.newArrayList());
++    }
++    public BlockShearEntityEvent(@NotNull Block dispenser, @NotNull Entity sheared, @NotNull ItemStack tool, @NotNull java.util.List<ItemStack> drops) {
++        // Paper end
+         super(dispenser);
+         this.sheared = sheared;
+         this.tool = tool;
++        this.drops = drops; // Paper
+     }
+ 
+     /**
+@@ -64,4 +72,25 @@ public class BlockShearEntityEvent extends BlockEvent implements Cancellable {
+     public static HandlerList getHandlerList() {
+         return handlers;
+     }
++    // Paper start
++    /**
++     * Get an immutable list of drops for this shearing.
```

Comment:
Immutable is not a guarantee. Perhaps document in an own block that its mutability is undefined, so still point towards `#setDrops`?

Reply:
Is the only reason it’s not guaranteed that someone could construct and fire this event in their own plugin with a mutable list?

Reply:
That and the fact you yourself pass in mutable lists quite a few places in the very same PR.

Reply:
Isn’t it just Lists.transform? I thought the docs on those said the resulting list is immutable.